src/GivensQR.jl

```diff
@@ -0,0 +1,141 @@
+module GivensQR
+
+export givensqr_op
+
+using SparseArrays, LinearAlgebra, LinearOperators
+
+mutable struct GivensQR_model{T}
+    nrows ::Int
+    ncols ::Int
+    rows ::Array{Int}
+    cols ::Array{Int}
+    cosines ::Array{T}
+    sines ::Array{T}
+    col_perm ::Array{Int}
+    rank ::Int
+end
+
+function qr(R :: SparseMatrixCSC{T,Ti}) where {Ti <: Integer, T}
+    m = R.m
+    n = R.n
+    rows = Int[]
+    cols = Int[]
+    cosines = T[]
+    sines = T[]
+    k = min(m,n)
+    no_null_col = n
+    col_perm = AbstractArray{Int}(1:n)
+    rank = 0
+
+    for j = 1:min(k, no_null_col)
+
+        while (no_null_col > j && R[j:m,j].nzind == [])
```

Comment:
It's possible to create a single function, by defining appropriate functions for this check. Such as
```julia
is_empty_under_pivot(M :: AbstractSparseMatrix) = ...
```